{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "a1559b58-7fd9-41c6-8421-900ebe2b9d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "\n",
    "def generate_genotype_matrix(n_snps: int):\n",
    "    \"\"\"\n",
    "    Generate a binary genotypes matrix given a number of total snps to model the genotypes of.\n",
    "    \n",
    "    Argument(s):\n",
    "    n_snps: number of total snps to model\n",
    "    \"\"\"\n",
    "    \n",
    "    #MAKE SURE THERE ARE ENOUGH SNPS TO GENERATE A GENOTYPE MATRIX\n",
    "    \n",
    "    #if there are less than 2 snps, only 1 conformation is possible\n",
    "    if n_snps<2:\n",
    "        #return the only possible conformation of no snps\n",
    "        return 0\n",
    "    \n",
    "    \n",
    "    #GENERATE PHENOTYPES AND HAPLOTYPES\n",
    "    \n",
    "    #number of phenotypes assuming 2 possible conformations for each snp\n",
    "    n_phenotypes = 2**n_snps\n",
    "\n",
    "    #all possible haplotype pairs\n",
    "    haplotypes = [list(p) for p in itertools.product(range(2), repeat=n_snps)]\n",
    "    #all possible combinations of phenotypes\n",
    "    expanded_grid = [(x, y) for x in range(n_phenotypes) for y in range(n_phenotypes)]\n",
    "    \n",
    "    \n",
    "    #GENERATE GENOTYPES\n",
    "\n",
    "    #initiate list to store genotypes\n",
    "    genotypes = []\n",
    "    #for each possible combination of (0 or 1)\n",
    "    for indices in expanded_grid:\n",
    "        #get associated haplotypes (binary)\n",
    "        haplotype_n = [haplotypes[i] for i in list(indices)]\n",
    "        #sum haplotypes to get genotype (0, 1, 2)\n",
    "        genotype_n = tuple(map(sum, zip(*haplotype_n)))\n",
    "        #append to master list\n",
    "        genotypes.append(genotype_n)\n",
    "    #make set of genotypes into strings for use as labels\n",
    "    genotypes = [('%snp'*len(x)%x).strip() for x in genotypes]\n",
    "\n",
    "    #all unique genotypes\n",
    "    unique_genotypes = [p for p in itertools.product(range(3), repeat=n_snps)]\n",
    "    unique_genotypes = [('%snp'*len(x)%x).strip() for x in unique_genotypes]\n",
    "\n",
    "    \n",
    "    #GENERATE GENOTYPE MATRIX\n",
    "    \n",
    "    #initiate genotype matrix with all 0s\n",
    "    genotype_matrix = np.zeros((len(unique_genotypes), n_phenotypes**2))\n",
    "    #for each unique genotype\n",
    "    for i in range(len(unique_genotypes)):\n",
    "        #find all locations within list of genotypes that corresponds to this genotype\n",
    "        locations = [j for j,d in enumerate(genotypes) if d==unique_genotypes[i]]\n",
    "        #impute with 1 if genotype label matches\n",
    "        genotype_matrix[i,locations] = 1\n",
    "\n",
    "        \n",
    "    #OUTPUT\n",
    "    \n",
    "    #return imputed genotype matrix\n",
    "    return genotype_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "ecf28dd2-04d9-42d1-a292-20d05f629a45",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "\n",
    "\n",
    "def generate_haplotype_probabilities(haplotypes_frequency: object,\n",
    "                                     fractional_frequency: object):\n",
    "    \"\"\"\n",
    "    Generate haplotype probabilities based on fractional frequencies.\n",
    "    \n",
    "    Argument(s):\n",
    "    haplotypes_frequency: matrix with rows as haplotypes and columns as snps\n",
    "    fractional_frequency: list with elements as fractional frequencies of each haplotype\n",
    "    \"\"\"\n",
    "\n",
    "    #GENERATE LABELS\n",
    "    \n",
    "    #number of haplotypes\n",
    "    n_haplotypes = haplotypes_frequency.shape[1]\n",
    "    #number of snps\n",
    "    n_snps = haplotypes_frequency.shape[0]\n",
    "    #number of possible binary combinations of haplotype pairs\n",
    "    n_haplotype_pairs = 2**n_haplotypes\n",
    "\n",
    "    #get labels as all possible haplotype configurations\n",
    "    labels = [list(p) for p in itertools.product(range(2), repeat=n_haplotypes)]\n",
    "    \n",
    "    \n",
    "    #GENERATE PROBABILITIES\n",
    "    \n",
    "    #initiate probabilties list\n",
    "    probabilities = [0]*n_haplotype_pairs\n",
    "    #initiate mapped probabilities dictionary\n",
    "    mapped_probability = {}\n",
    "\n",
    "    #for each haplotype pair\n",
    "    for i in range(n_haplotype_pairs):\n",
    "        #initiate probability map with haplotype labels as indexes\n",
    "        mapped_probability[str(labels[i])] = 0\n",
    "\n",
    "    #for each snp\n",
    "    for i in range(n_snps):\n",
    "        #initiate blank label\n",
    "        line = [0]*n_haplotypes\n",
    "        #for each haplotype\n",
    "        for j in range(n_haplotypes):\n",
    "            #if this is the correct haplotype\n",
    "            if haplotypes_frequency.iloc[i, j] != 0:\n",
    "                #indicate this label is present for this snp\n",
    "                line[j] = 1\n",
    "        #the haplotype probability is the cummulative sum of the fractional frequency from all haplotypes associated with this snp\n",
    "        mapped_probability[str(line)] = mapped_probability[str(line)] + fractional_frequency[i]\n",
    "\n",
    "    #for each haplotype pair\n",
    "    for i in range(n_haplotype_pairs):\n",
    "        #record haplotype probabilities to probabilities vector\n",
    "        probabilities[i] = mapped_probability[str(labels[i])]\n",
    "    \n",
    "    \n",
    "    #OUTPUT\n",
    "    \n",
    "    #return probabilities vector\n",
    "    return probabilities\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "05cdde41-5724-4778-8ce7-d3fb8c3edaec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "\n",
    "\n",
    "def generate_genotype_probabilities(genotypes: object,\n",
    "                                    probabilities: list):\n",
    "    \"\"\"\n",
    "    Generate genotype probabilities from genotypes and haplotype probabilities.\n",
    "    \n",
    "    Argument(s):\n",
    "    genotypes: binary matrix with rows as unique genotypes and columns as all possible genotypes\n",
    "    probabilities: list with elements as probabilities of each haplotype pair\n",
    "    \"\"\"\n",
    "    \n",
    "    #COUNT POSSIBLE GENOTYPE\n",
    "    \n",
    "    #number of unique genotypes\n",
    "    n_unique_genotypes = genotypes.shape[0]\n",
    "    #total number of possible genotypes\n",
    "    n_total_genotypes = genotypes.shape[1]\n",
    "    #initiate output list to store genotype probabilities\n",
    "    genotype_probabilities = [0]*n_unique_genotypes\n",
    "\n",
    "    \n",
    "    #CALCULATE GENOTYPES PROBABILITIES\n",
    "        \n",
    "    #for each possible genotype\n",
    "    for i in range(n_total_genotypes):\n",
    "        #get reference indicating which unique genotype this genotype matches\n",
    "        reference = genotypes[:, i]\n",
    "        #probability of this genotype\n",
    "        probability_n = probabilities[i]\n",
    "\n",
    "        #initiate loop variable\n",
    "        j = 0\n",
    "        #while j is in the neighborhood of the snp\n",
    "        while j < math.floor(n_unique_genotypes/4)*4:\n",
    "            #generate initial probabilities for the snp as well as its neighbor\n",
    "            v0_initial = reference[j] * probability_n\n",
    "            v1_initial = reference[j+2] * probability_n\n",
    "            #add reference status to probability\n",
    "            v0 = reference[j] + v0_initial\n",
    "            v1 = reference[j+2] + v1_initial\n",
    "            #input into genotype probabilities vector\n",
    "            genotype_probabilities[j] = v0\n",
    "            genotype_probabilities[j+2] = v1\n",
    "            #stride of 2 to bypass neighbor with effect already generated\n",
    "            j = j+4\n",
    "        #for the last genotype with no weighbor\n",
    "        while j < n_unique_genotypes:\n",
    "            #probability is equal to neighborhood effect from past genotype along with calculated standard probability\n",
    "            genotype_probabilities[j] = genotype_probabilities[j] + reference[j]*probabilities[i]\n",
    "            j = j+1\n",
    "\n",
    "            \n",
    "    #OUTPUT\n",
    "    \n",
    "    #return vector of probabilities for each genotype\n",
    "    return genotype_probabilities\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "145c8ffa-e440-431d-b2cd-620d5a368771",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "def compute_snp_probabilities(snp: str,\n",
    "                              causal_variants: list,\n",
    "                              haplotypes_data: list):\n",
    "    \"\"\"\n",
    "    Compute genotype conditional probabilities for each SNP.\n",
    "    \n",
    "    Argument(s):\n",
    "    snp: the name of the snp\n",
    "    causal_variants: a list of all causal variant names\n",
    "    haplotypes_data: matrix with rows as haplotypes and columns as snps, with a final column as fractional frequencies of each haplotype\n",
    "    \"\"\"\n",
    "    \n",
    "    #GENERATE GENOTYPE MATRICES\n",
    "    \n",
    "    #generate all possible genotype configurations matrix\n",
    "    geno_matrix_2snps = generate_genotype_matrix(2)\n",
    "    geno_matrix_3snps = generate_genotype_matrix(3)\n",
    "    geno_matrix_4snps = generate_genotype_matrix(4)\n",
    "    geno_matrix_5snps = generate_genotype_matrix(5)\n",
    "    geno_matrix_6snps = generate_genotype_matrix(6)\n",
    "\n",
    "    #generate all possible combinations of snp alleles given the number of snps\n",
    "    geno_2SNP = [list(p) for p in itertools.product(range(3), repeat=2)]\n",
    "    geno_3SNP = [list(p) for p in itertools.product(range(3), repeat=3)]\n",
    "    geno_4SNP = [list(p) for p in itertools.product(range(3), repeat=4)]\n",
    "    geno_5SNP = [list(p) for p in itertools.product(range(3), repeat=5)]\n",
    "    geno_6SNP = [list(p) for p in itertools.product(range(3), repeat=6)]\n",
    "\n",
    "    #find all indices where the first snp is a specific allele\n",
    "    which_X0_2SNP = [index for index, num_list in enumerate(geno_2SNP) if num_list[0] == 0]\n",
    "    which_X1_2SNP = [index for index, num_list in enumerate(geno_2SNP) if num_list[0] == 1]\n",
    "    which_X2_2SNP = [index for index, num_list in enumerate(geno_2SNP) if num_list[0] == 2]\n",
    "\n",
    "    which_X0_3SNP = [index for index, num_list in enumerate(geno_3SNP) if num_list[0] == 0]\n",
    "    which_X1_3SNP = [index for index, num_list in enumerate(geno_3SNP) if num_list[0] == 1]\n",
    "    which_X2_3SNP = [index for index, num_list in enumerate(geno_3SNP) if num_list[0] == 2]\n",
    "\n",
    "    which_X0_4SNP = [index for index, num_list in enumerate(geno_4SNP) if num_list[0] == 0]\n",
    "    which_X1_4SNP = [index for index, num_list in enumerate(geno_4SNP) if num_list[0] == 1]\n",
    "    which_X2_4SNP = [index for index, num_list in enumerate(geno_4SNP) if num_list[0] == 2]\n",
    "\n",
    "    which_X0_5SNP = [index for index, num_list in enumerate(geno_5SNP) if num_list[0] == 0]\n",
    "    which_X1_5SNP = [index for index, num_list in enumerate(geno_5SNP) if num_list[0] == 1]\n",
    "    which_X2_5SNP = [index for index, num_list in enumerate(geno_5SNP) if num_list[0] == 2]\n",
    "\n",
    "    which_X0_6SNP = [index for index, num_list in enumerate(geno_6SNP) if num_list[0] == 0]\n",
    "    which_X1_6SNP = [index for index, num_list in enumerate(geno_6SNP) if num_list[0] == 1]\n",
    "    which_X2_6SNP = [index for index, num_list in enumerate(geno_6SNP) if num_list[0] == 2]\n",
    "    \n",
    "\n",
    "    #FIND CONTRIBUTING SNPS\n",
    "    \n",
    "    #if the snp is a causal variant\n",
    "    if snp in causal_variants:\n",
    "        #the list of contributing snps are all causal variants\n",
    "        snp_list = causal_variants\n",
    "    #if the snp is not a causal variant\n",
    "    else:\n",
    "        #the list of contributing snps is this snp along with all causal variants\n",
    "        snp_list = [snp]+causal_variants\n",
    "        \n",
    "    #number of snps contributing to this snps effect\n",
    "    n_snps = len(snp_list)\n",
    "    \n",
    "    #the frequency matrix is the haplotype frequency list for all contributing snps\n",
    "    haplotypes_frequency = haplotypes_data[snp_list].sub(1)\n",
    "    #the probability vector is the probability of all haplotypes\n",
    "    fractional_frequency = haplotypes_data['probability']\n",
    "    \n",
    "    #calculate haplotype probabilities\n",
    "    haplotypes_probability = np.array(generate_haplotype_probabilities(haplotypes_frequency,fractional_frequency))\n",
    "    \n",
    "    \n",
    "    #GENERATE CONDITIONAL PROBABILITIES\n",
    "\n",
    "    #if there is only 1 contributing snp\n",
    "    if n_snps == 1:\n",
    "        #if snp in causal_variants:\n",
    "        #the effect at this snp follows the equasion a^2+2ab+b^2\n",
    "        function_output = [[haplotypes_probability[0]**2, 0, 0],\n",
    "                           [0, 2*haplotypes_probability[0]*haplotypes_probability[1], 0],\n",
    "                           [0, 0, haplotypes_probability[1]**2]]\n",
    "        # else: #delete_me???\n",
    "        # #the effect at this snp follows the equasion a^2+2ab+b^2\n",
    "        #     function_output = [[haplotypes_probability[1]**2, 0, 0],\n",
    "        #                        [haplotypes_probability[1]**2, 0, 0],\n",
    "        #                        [haplotypes_probability[1]**2, 0, 0]]\n",
    "\n",
    "    #if there are 2 contributing snps\n",
    "    elif n_snps == 2:\n",
    "        \n",
    "        #outer product of haplotype probabilities vector to get probabilities of all paired haplotypes\n",
    "        vector_input = [x for xs in np.outer(haplotypes_probability, haplotypes_probability.T).tolist() for x in xs]\n",
    "        #generate genotype probabilities\n",
    "        genotype_probabilities = generate_genotype_probabilities(geno_matrix_2snps, vector_input)\n",
    "        \n",
    "        #if the snp is a causal variant\n",
    "        if snp in causal_variants:\n",
    "            #find the index of the snp in the list of all contributing snps\n",
    "            snp_index = causal_variants.index(snp)\n",
    "            #get snp specific genotype information for each genotype\n",
    "            snp_specific = [i[snp_index] for i in geno_2SNP]\n",
    "            #calculate conditional probabilities based on genotype probabilities multiplied by snp specific effect\n",
    "            function_output =  [[a*b for a,b in zip(genotype_probabilities,[int(i == 0) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 1) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 2) for i in snp_specific])]]\n",
    "        #if the snp is not a causal variant\n",
    "        else:\n",
    "            #get conditional probabilities directly from genotype probabilities\n",
    "            function_output =  [[genotype_probabilities[i] for i in which_X0_2SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X1_2SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X2_2SNP]]\n",
    "\n",
    "    #if there are 3 contributing snps\n",
    "    elif n_snps == 3:\n",
    "        \n",
    "        #outer product of haplotype probabilities vector to get probabilities of all paired haplotypes\n",
    "        vector_input = [x for xs in np.outer(haplotypes_probability, haplotypes_probability.T).tolist() for x in xs]\n",
    "        #generate genotype probabilities\n",
    "        genotype_probabilities = generate_genotype_probabilities(geno_matrix_3snps, vector_input)\n",
    "        \n",
    "        #if the snp is a causal variant\n",
    "        if snp in causal_variants:\n",
    "            #find the index of the snp in the list of all contributing snps\n",
    "            snp_index = causal_variants.index(snp)\n",
    "            #get snp specific genotype information for each genotype\n",
    "            snp_specific = [i[snp_index] for i in geno_3SNP]\n",
    "            #calculate conditional probabilities based on genotype probabilities multiplied by snp specific effect\n",
    "            function_output =  [[a*b for a,b in zip(genotype_probabilities,[int(i == 0) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 1) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 2) for i in snp_specific])]]\n",
    "        \n",
    "        #if the snp is not a causal variant\n",
    "        else:\n",
    "            #get conditional probabilities directly from genotype probabilities\n",
    "            function_output =  [[genotype_probabilities[i] for i in which_X0_3SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X1_3SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X2_3SNP]]\n",
    "            \n",
    "    #if there are 4 contributing snps\n",
    "    elif n_snps == 4:\n",
    "        \n",
    "        #outer product of haplotype probabilities vector to get probabilities of all paired haplotypes\n",
    "        vector_input = [x for xs in np.outer(haplotypes_probability, haplotypes_probability.T).tolist() for x in xs]\n",
    "        #generate genotype probabilities\n",
    "        genotype_probabilities = generate_genotype_probabilities(geno_matrix_4snps, vector_input)\n",
    "        \n",
    "        #if the snp is a causal variant\n",
    "        if snp in causal_variants:\n",
    "            #find the index of the snp in the list of all contributing snps\n",
    "            snp_index = causal_variants.index(snp)\n",
    "            #get snp specific genotype information for each genotype\n",
    "            snp_specific = [i[snp_index] for i in geno_4SNP]\n",
    "            #calculate conditional probabilities based on genotype probabilities multiplied by snp specific effect\n",
    "            function_output =  [[a*b for a,b in zip(genotype_probabilities,[int(i == 0) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 1) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 2) for i in snp_specific])]]\n",
    "        \n",
    "        #if the snp is not a causal variant\n",
    "        else:\n",
    "            #get conditional probabilities directly from genotype probabilities\n",
    "            function_output =  [[genotype_probabilities[i] for i in which_X0_4SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X1_4SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X2_4SNP]]\n",
    "            \n",
    "    #if there are 5 contributing snps\n",
    "    elif n_snps == 5:\n",
    "        \n",
    "        #outer product of haplotype probabilities vector to get probabilities of all paired haplotypes\n",
    "        vector_input = [x for xs in np.outer(haplotypes_probability, haplotypes_probability.T).tolist() for x in xs]\n",
    "        #generate genotype probabilities\n",
    "        genotype_probabilities = generate_genotype_probabilities(geno_matrix_5snps, vector_input)\n",
    "        \n",
    "        #if the snp is a causal variant\n",
    "        if snp in causal_variants:\n",
    "            #find the index of the snp in the list of all contributing snps\n",
    "            snp_index = causal_variants.index(snp)\n",
    "            #get snp specific genotype information for each genotype\n",
    "            snp_specific = [i[snp_index] for i in geno_5SNP]\n",
    "            #calculate conditional probabilities based on genotype probabilities multiplied by snp specific effect\n",
    "            function_output =  [[a*b for a,b in zip(genotype_probabilities,[int(i == 0) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 1) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 2) for i in snp_specific])]]\n",
    "        \n",
    "        #if the snp is not a causal variant\n",
    "        else:\n",
    "            #get conditional probabilities directly from genotype probabilities\n",
    "            function_output =  [[genotype_probabilities[i] for i in which_X0_5SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X1_5SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X2_5SNP]]\n",
    "            \n",
    "    #if there are 6 contributing snps\n",
    "    elif n_snps == 6:\n",
    "        \n",
    "        #outer product of haplotype probabilities vector to get probabilities of all paired haplotypes\n",
    "        vector_input = [x for xs in np.outer(haplotypes_probability, haplotypes_probability.T).tolist() for x in xs]\n",
    "        #generate genotype probabilities\n",
    "        genotype_probabilities = generate_genotype_probabilities(geno_matrix_6snps, vector_input)\n",
    "        \n",
    "        #if the snp is a causal variant\n",
    "        if snp in causal_variants:\n",
    "            #find the index of the snp in the list of all contributing snps\n",
    "            snp_index = causal_variants.index(snp)\n",
    "            #get snp specific genotype information for each genotype\n",
    "            snp_specific = [i[snp_index] for i in geno_6SNP]\n",
    "            #calculate conditional probabilities based on genotype probabilities multiplied by snp specific effect\n",
    "            function_output =  [[a*b for a,b in zip(genotype_probabilities,[int(i == 0) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 1) for i in snp_specific])],\n",
    "                                [a*b for a,b in zip(genotype_probabilities,[int(i == 2) for i in snp_specific])]]\n",
    "        \n",
    "        #if the snp is not a causal variant\n",
    "        else:\n",
    "            #get conditional probabilities directly from genotype probabilities\n",
    "            function_output =  [[genotype_probabilities[i] for i in which_X0_6SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X1_6SNP],\n",
    "                                [genotype_probabilities[i] for i in which_X2_6SNP]]   \n",
    "            \n",
    "    #if more than 6 contributing snps\n",
    "    else:\n",
    "        #report error\n",
    "        print('too many snps; >6 total snps')\n",
    "        \n",
    "        \n",
    "    #OUTPUT\n",
    "\n",
    "    #return vector of conditional probabilities\n",
    "    return function_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "d2371ca6-1f9f-4c55-89b5-650493f438e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import math\n",
    "\n",
    "\n",
    "def compute_initial_gamma(n_controls: int,\n",
    "                          n_cases: int,\n",
    "                          causal_variants: list,\n",
    "                          log_odds_ratios: list,\n",
    "                          haplotypes_data: list):\n",
    "    \"\"\"\n",
    "    Compute initial gamma values.\n",
    "    \n",
    "    Argument(s):\n",
    "    n_controls: number of controls\n",
    "    n_cases: number of cases\n",
    "    causal_variants: names of all causal variants\n",
    "    log_odds_ratios: log odds ratios for all causal variants\n",
    "    haplotypes_data: matrix with rows as haplotypes and columns as snps, with a final column as fractional frequencies of each haplotype    \n",
    "    \"\"\"\n",
    "\n",
    "    input_to_sweep = [list(p) for p in itertools.product(range(3), repeat=len(causal_variants))]\n",
    "    input_to_sweep = pd.DataFrame(input_to_sweep)\n",
    "    input_to_sweep = input_to_sweep.apply(lambda row: row * log_odds_ratios, axis=1)\n",
    "    eta_CV = input_to_sweep.sum(axis=1)\n",
    "\n",
    "    e_eta_CV = math.e**eta_CV\n",
    "\n",
    "    GenoProbW = compute_snp_probabilities(causal_variants[0], causal_variants, haplotypes_data)\n",
    "    \n",
    "    PW = [sum(x) for x in zip(*GenoProbW)]\n",
    "\n",
    "    denom = e_eta_CV*PW\n",
    "    \n",
    "    if sum(denom) == 0:\n",
    "        initial_gamma = math.log(n_cases)-math.log(n_controls)\n",
    "    else:\n",
    "        initial_gamma = math.log(n_cases)-math.log(n_controls)-math.log(sum(denom))\n",
    "\n",
    "    return initial_gamma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "1d4f1610-9b7f-46ad-80da-9ff3cf94993f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import math\n",
    "import itertools\n",
    "\n",
    "def estimate_z_scores(n_controls: int,\n",
    "                        n_cases: int,\n",
    "                        snps: list,\n",
    "                        causal_variants: list,\n",
    "                        log_odds_ratios: list,\n",
    "                        haplotypes_data: list,\n",
    "                        genotype_probabilities: list):\n",
    "\n",
    "    \"\"\"\n",
    "    Estimate z-scores for all SNPs.\n",
    "    \n",
    "    Argument(s):\n",
    "    n_controls: number of controls\n",
    "    n_cases: number of cases\n",
    "    snps: names of all SNPs\n",
    "    causal_variants: names of all causal variants\n",
    "    log_odds_ratios: log odds ratios for all causal variants\n",
    "    haplotypes_data: matrix with rows as haplotypes and columns as snps, with a final column as fractional frequencies of each haplotype \n",
    "    genotype_probabilities: a list of lists of all genotype conditional probabilities for each SNP\n",
    "    \"\"\"\n",
    "    \n",
    "    #INITIATE CONSTANTS\n",
    "    \n",
    "    #compute inital gamma value\n",
    "    initial_gamma = compute_initial_gamma(n_controls, n_cases, causal_variants, log_odds_ratios, haplotypes_data)\n",
    "    #total number of individuals\n",
    "    n_total = n_controls + n_cases\n",
    "\n",
    "\n",
    "    #COMPUTE INPUTS\n",
    "    \n",
    "    #compute conditional probability of being causal\n",
    "    input_to_sweep = [list(p) for p in itertools.product(range(3), repeat=len(causal_variants))]\n",
    "    input_to_sweep = pd.DataFrame(input_to_sweep)\n",
    "    input_to_sweep = input_to_sweep.apply(lambda row: row * log_odds_ratios, axis=1)\n",
    "    eta_CV = input_to_sweep.sum(axis=1)\n",
    "    expeta = math.e ** (initial_gamma + eta_CV)\n",
    "    \n",
    "    #compute multiplicative factors\n",
    "    u_factor = n_controls * (n_total - 1) * (n_controls * expeta - n_cases) / (n_total ** 2)\n",
    "    power_factor = n_controls * (expeta + 1) / n_total\n",
    "\n",
    "    \n",
    "    #ESTIMATE Z SCORES\n",
    "    \n",
    "    #initiate output list\n",
    "    z_scores = []\n",
    "    #for each snp\n",
    "    for i in range(len(snps)):\n",
    "        #estimate the snps z-score\n",
    "        z_score_n = estimate_z_score_n(n_controls, n_cases, u_factor, power_factor, genotype_probabilities[i])\n",
    "        #append the z score to the output list\n",
    "        z_scores.append(z_score_n)\n",
    "    \n",
    "    \n",
    "    #OUTPUT\n",
    "    \n",
    "    #return list of estimated z-scores for each snp\n",
    "    return z_scores\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "c6a8fd90-903b-4e67-bb12-c8c241a5c020",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import math\n",
    "\n",
    "\n",
    "def estimate_z_score_n(n_controls: int,\n",
    "                     n_cases: int,\n",
    "                     u_factor: float,\n",
    "                     power_factor: float,\n",
    "                     genotype_probability_n: list):\n",
    "\n",
    "    \"\"\"\n",
    "    Estimate the z-score for a single SNP.\n",
    "    \n",
    "    Argument(s):\n",
    "    n_controls: number of controls\n",
    "    n_cases: number of cases\n",
    "    u_factor: multiplicative factor\n",
    "    power_factor: multiplicative factor\n",
    "    genotype_probabilities: a list of lists of all genotype conditional probabilities for each SNP\n",
    "    \"\"\"\n",
    "    \n",
    "    PX1W = genotype_probability_n[1]\n",
    "    PX2W = genotype_probability_n[2]\n",
    "\n",
    "    n_total = n_controls + n_cases\n",
    "    u_factorPX2W = sum(u_factor * PX2W)\n",
    "    u_factorPX1W = sum(u_factor * PX1W)\n",
    "    power_factorPX2W = sum(power_factor * PX2W)\n",
    "    power_factorPX1W = sum(power_factor * PX1W)\n",
    "    U = 2 * u_factorPX2W + u_factorPX1W\n",
    "\n",
    "    EX = 2 * power_factorPX2W + power_factorPX1W\n",
    "    EX2 = 4 * power_factorPX2W + power_factorPX1W\n",
    "    EX3 = 8 * power_factorPX2W + power_factorPX1W\n",
    "    EX4 = 16 * power_factorPX2W + power_factorPX1W\n",
    "\n",
    "    EVX = EX2 - (EX ** 2.0)\n",
    "\n",
    "    Term1 = n_total * EX4 + n_total * (n_total - 1) * (EX2 ** 2.0)\n",
    "    Term2 = EX4+2*(n_total-1)*EX3*EX+(n_total-1)*(EX2**2.0)+(n_total-1)*(n_total-2)*EX2*(EX**2.0)\n",
    "    Term3 = (EX4+4*(n_total-1)*EX3*EX+6*(n_total-1)*(EX2**2.0)+6*(n_total-1)*(n_total-2)*EX2*(EX**2.0)+(n_total-1)*(n_total-2)*(n_total-3)*(EX**4.0))/n_total\n",
    "    EVX2 = ((1/(n_total-1))**2.0) * (Term1 - (2*Term2) + (Term3))\n",
    "\n",
    "    a = np.float64((2*EVX2-EVX*EVX))/(EVX2-EVX*EVX)\n",
    "    b = np.float64((EVX2*EVX))/(EVX2-EVX*EVX)\n",
    "\n",
    "    EinvsqrtVX = (b**(-0.5))*math.e**(math.lgamma((2*a+1)/2)-math.lgamma(a))\n",
    "    z_score_n = U*EinvsqrtVX*((n_total/(n_controls*n_cases))**0.5)\n",
    "\n",
    "    return z_score_n\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "5f3c9126-9601-4c7f-b879-2be2f4c0ca25",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "\n",
    "\n",
    "def simulate_p_values(n_snps: int,\n",
    "                      n_haplotypes: int,\n",
    "                      neighborhood_length: int,\n",
    "                      maf: list,\n",
    "                      causal_variants: list,\n",
    "                      odds_ratios: list,\n",
    "                      n_controls: int,\n",
    "                      n_cases: int):\n",
    "\n",
    "    \"\"\"\n",
    "    Generate a set of p-values.\n",
    "    \n",
    "    Argument(s):\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    n_controls: number of controls\n",
    "    n_cases: number of cases\n",
    "    \"\"\"\n",
    "\n",
    "    #GENERATE HAPLOTYPES\n",
    "    \n",
    "    #initiate haplotypes with length including neighborhood\n",
    "    haplotypes_neighborhood = np.empty((n_haplotypes, n_snps + neighborhood_length))\n",
    "\n",
    "    #for each snp, plus snps of the boundary neighborhood\n",
    "    for i in range(n_snps+neighborhood_length):\n",
    "        #generate random alleles\n",
    "        haplotypes_neighborhood[:, i] = np.random.binomial(1, maf[i], size=n_haplotypes)\n",
    "    #crop neighborhood off\n",
    "    haplotypes = haplotypes_neighborhood[: ,0:n_snps]\n",
    "    \n",
    "    #for the neighborhood of each snp\n",
    "    for j in range(neighborhood_length):\n",
    "        #generate individual neirghborhoods for each snp\n",
    "        input_locations = [x+j for x in range(n_snps)]\n",
    "        #redefine haplotypes to include neighborhood effects\n",
    "        haplotypes = haplotypes + haplotypes_neighborhood[:, input_locations]\n",
    "\n",
    "\n",
    "    #GENERATE HAPLOTYPE FREQUENCY MATRIX\n",
    "    \n",
    "    #initiate haplotype frequency matrix\n",
    "    temp = np.empty((n_haplotypes,n_snps))\n",
    "    #for each snp\n",
    "    for i in range(n_snps):\n",
    "        #take the maximum haplotype value of that snp\n",
    "        max_temp = max(haplotypes[:, i])\n",
    "        temp[:,i] = max_temp\n",
    "    #haplotype prequency matrix is teh haplotypes divided by the maximum allele value\n",
    "    haplotypes = np.round(np.divide(haplotypes, temp))\n",
    "\n",
    "    #name of all snps\n",
    "    snps = ['snp' + str(num1) for num1 in range(n_snps)]\n",
    "    #turn from 0,1 to 1,2\n",
    "    new_haplotypes = haplotypes+1\n",
    "    #create haplotype matrix\n",
    "    frequency_matrix = pd.DataFrame(data = new_haplotypes, columns = snps)\n",
    "    #equal probabilities for all haplotypes\n",
    "    frequency_matrix['probability'] = 1/len(frequency_matrix)\n",
    "\n",
    "\n",
    "    #CALCULATE CONDITIONAL GENOTYPE PROBABILITIES\n",
    "\n",
    "    #initiate list to store vectors of conditional genotype probabilities\n",
    "    genotype_probabilities = []\n",
    "    #for each snp\n",
    "    for snp in snps:\n",
    "        #compute conditional probabilities given an snp\n",
    "        snp_output = compute_snp_probabilities(snp, causal_variants, frequency_matrix)\n",
    "        #append to all conditional probailities\n",
    "        genotype_probabilities.append(snp_output)\n",
    "\n",
    "    \n",
    "    #ESTIMATE Z-SCORES\n",
    "    \n",
    "    #calculate log odds ratio\n",
    "    log_odds_ratios = [math.log(i) for i in odds_ratios]\n",
    "    #estimate z scores\n",
    "    exp_z_score = estimate_z_scores(n_controls, n_cases, snps, causal_variants, log_odds_ratios, frequency_matrix, genotype_probabilities)\n",
    "    print('number of nan z scores: ', len([0 for x in exp_z_score if math.isnan(x)]))\n",
    "    \n",
    "    #CALCULATE P-VALUES\n",
    "    \n",
    "    #initiate list to store p values\n",
    "    p_values = []\n",
    "    #for each z score\n",
    "    for i in range(len(exp_z_score)):\n",
    "        #calculate p value from z score\n",
    "        p_value_n = scipy.stats.norm.sf(abs(exp_z_score[i]))\n",
    "        #append to list of p values\n",
    "        p_values.append(p_value_n)\n",
    "        \n",
    "        \n",
    "    #OUTPUT\n",
    "    \n",
    "    #list of p values\n",
    "    return p_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "1512151b-7fb4-42ec-96fd-afd6d0c50e3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#==========================================================================================================================================="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3935d3c7-ceb8-4a2e-bcca-2ca627067821",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import random\n",
    "\n",
    "# n_snps = 1882\n",
    "# n_haplotypes = 100\n",
    "# neighborhood_length = 5\n",
    "# maf = [random.uniform(0.05, 0.5) for _ in range(n_snps+neighborhood_length)]\n",
    "# causal_variants = ['snp599', 'snp1333']\n",
    "# odds_ratios = [4.8608503825977638, 4.5139286248169914]\n",
    "\n",
    "# n_controls = 50\n",
    "# n_cases = 50\n",
    "    \n",
    "    \n",
    "# import time\n",
    "# start_time = time.time()\n",
    "\n",
    "# p_values = simulate_p_values(n_snps, n_haplotypes, neighborhood_length, maf, causal_variants, odds_ratios, n_controls, n_cases)\n",
    "\n",
    "# end_time = time.time()\n",
    "# time_elapsed = end_time-start_time\n",
    "# time_elapsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e7dab5-42a7-4533-a56f-8b27fcd5e703",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #plot p values\n",
    "# import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "# neg_log_p_values = -1*np.log10(p_values)\n",
    "\n",
    "\n",
    "# plt.scatter(range(len(neg_log_p_values)), neg_log_p_values)\n",
    "# vert_line_pos = [i for i, e in enumerate(['snp' + str(num1) for num1 in range(n_snps)]) if e in causal_variants]\n",
    "# for position in vert_line_pos:\n",
    "#     plt.axvline(x = position, color = 'b', label = 'axvline - full height')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca897a88-476f-4fc8-b7bd-80c020f204a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import random\n",
    "\n",
    "# n_snps = 1000\n",
    "# n_haplotypes = 1000\n",
    "# neighborhood_length = 5\n",
    "# maf = [random.uniform(0.4, 0.5) for _ in range(n_snps+neighborhood_length)]\n",
    "# causal_variants = ['snp59', 'snp88']\n",
    "# odds_ratios = [1.8608503825977638, 3.5139286248169914]\n",
    "\n",
    "# n_controls = 1000\n",
    "# n_cases = 1000\n",
    "    \n",
    "    \n",
    "# import time\n",
    "# start_time = time.time()\n",
    "\n",
    "# p_values = simulate_p_values(n_snps, n_haplotypes, neighborhood_length, maf, causal_variants, odds_ratios, n_controls, n_cases)\n",
    "\n",
    "# end_time = time.time()\n",
    "# time_elapsed = end_time-start_time\n",
    "# time_elapsed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7b1f4a8-7e65-4c5f-9dfe-b1f5460dcd0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #plot p values\n",
    "# import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "# neg_log_p_values = -1*np.log10(p_values)\n",
    "\n",
    "\n",
    "# plt.scatter(range(len(neg_log_p_values)), neg_log_p_values)\n",
    "# vert_line_pos = [i for i, e in enumerate(['snp' + str(num1) for num1 in range(n_snps)]) if e in causal_variants]\n",
    "# for position in vert_line_pos:\n",
    "#     plt.axvline(x = position, color = 'b', label = 'axvline - full height')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "7f75e3ab-3cec-4b65-bf47-e1837967f5dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#=============================================================================="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6b36de4e-3dc9-477a-8fcf-c4397a0bb590",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import random\n",
    "import scipy.stats as stats\n",
    "import math\n",
    "\n",
    "random.seed(25252)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e2cd0a56-954d-4b4b-b358-d470a0311676",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ANNOTATIONS (OUTPUT = \"annotations\")\n",
    "\n",
    "#import gencode annottions\n",
    "gencode = pd.read_table(\"gencode.v19.chr_patch_hapl_scaff.annotation.gtf\", comment=\"#\",\n",
    "                        sep = \"\\t\", names = ['seqname', 'source', 'feature', 'start' , 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "\n",
    "#limit to only these functional regions (do not want gene as that double counts the data; don't know what Selenocysteine is)\n",
    "gencode = gencode[gencode[\"feature\"].isin(['exon', 'CDS', 'start_codon', 'stop_codon', 'UTR'])]\n",
    "\n",
    "#limit to only chr1-21 and x,y,m\n",
    "gencode = gencode[gencode[\"seqname\"].isin(['chr1'])]\n",
    "\n",
    "#extract gene naames, type, and known/novel/putative status\n",
    "def gene_info(x):\n",
    "    # Extract gene names\n",
    "    g_name = list(filter(lambda x: 'gene_name' in x,  x.split(\"; \")))[0].split(\" \")[1]\n",
    "    g_type = list(filter(lambda x: 'gene_type' in x,  x.split(\"; \")))[0].split(\" \")[1]\n",
    "    g_status = list(filter(lambda x: 'gene_status' in x,  x.split(\"; \")))[0].split(\" \")[1]\n",
    "    return (g_name, g_type, g_status)\n",
    "gencode[\"gene_name\"], gencode[\"gene_type\"], gencode[\"gene_status\"] = zip(*gencode.attribute.apply(lambda x: gene_info(x)))\n",
    "\n",
    "\n",
    "#limit to only these columns for final dataframe\n",
    "gencode = gencode[['seqname', 'feature', 'start', 'end', 'strand', 'gene_name', 'gene_type', 'gene_status']]\n",
    "\n",
    "annotations = gencode\n",
    "\n",
    "\n",
    "annotations.columns = ['chromosome', 'annotation', 'start', 'end', 'gene_strand', 'gene_name', 'gene_type', 'gene_status']\n",
    "#only want the integer part; drop the \"chr\"\n",
    "annotations['chromosome'] = annotations['chromosome'].astype('str').str.extractall('(\\d+)').unstack().fillna('').sum(axis=1).astype(int)\n",
    "\n",
    "annotations = annotations.replace('\"', '', regex=True).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "aa935b61-52e2-4937-b91d-f9ebcbc9fdbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#CONDENSE_ANNOTATIONS\n",
    "\n",
    "def condense_annotations(annotations):\n",
    "    \"\"\"\n",
    "    Argument(s):\n",
    "    annotations: a dataframe with columns <chromosome, annotation, start, end, gene_{feature; e.g., name, strand, type}>\n",
    "    \"\"\"\n",
    "\n",
    "    #MERGE ANNOTATIONS\n",
    "\n",
    "    #columns in annotaion dataframe\n",
    "    annotation_columns = list(annotations.columns.values)\n",
    "    #remove start and end column names\n",
    "    annotation_columns_no_positions = [e for e in annotation_columns if e not in ('start', 'end')]\n",
    "    #group annotations dataframe by all gene features (all column names besides start and end) and find minimum start and maximum end value, condense all groups into one row with min start and max end values\n",
    "    annotations_out = annotations.groupby(annotation_columns_no_positions).agg({'start':'min', 'end':'max'})\n",
    "    #sort by chromosme first then start position\n",
    "    annotations_out = annotations_out.sort_values(['chromosome', 'start'], ascending=[True, True]).reset_index()\n",
    "\n",
    "\n",
    "    return annotations_out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b8332ed6-e0db-4c5a-9da3-f52e2f475d04",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations = condense_annotations(annotations)\n",
    "\n",
    "\n",
    "ld_blocks = pd.read_csv('fourier_ls-all.txt', sep=\"\t\")\n",
    "ld_blocks.columns = ['chromosome', 'start', 'end']\n",
    "ld_blocks['chromosome'] = ld_blocks['chromosome'].astype('str').str.extractall('(\\d+)').unstack().fillna('').sum(axis=1)\n",
    "ld_blocks = ld_blocks[ld_blocks['chromosome'].isin([1])].reset_index(drop=True)\n",
    "\n",
    "\n",
    "# annotations = annotations[0:800].reset_index(drop=True)\n",
    "# ld_blocks = ld_blocks[ld_blocks['end'] < annotations.loc[len(annotations)-1, 'end']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25c2845c-4728-4665-a11c-0b039680f595",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f57eb180-6896-42f5-b54c-9f062cf1229c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#===================================================================================================================="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "70d5d124-a30f-47bb-955f-d10271cd5aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations['start'] = annotations['start'].div(1000).round(0)\n",
    "annotations['end'] = annotations['end'].div(1000).round(0)\n",
    "\n",
    "ld_blocks['start'] = ld_blocks['start'].div(1000).round(0)\n",
    "ld_blocks['end'] = ld_blocks['end'].div(1000).round(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d6502e1f-c3e1-4627-b949-397047c254b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations.loc[:,\"end\"] = annotations[\"end\"] + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1f40f623-6f80-4473-b1f7-c4712384c879",
   "metadata": {},
   "outputs": [],
   "source": [
    "ld_blocks = ld_blocks[0:10].reset_index(drop=True)\n",
    "annotations = annotations[annotations['end'] < ld_blocks.loc[len(ld_blocks)-1, 'end']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "aec2fd78-184e-4491-8228-d1eb5b37989b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#======"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "271958fd-cff7-46f7-ab4f-7ad204a44f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(25252)\n",
    "\n",
    "\n",
    "#list of all genes (output: genes_list)\n",
    "genes_list = list(annotations['gene_name'].unique())\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "07313d0c-5cc4-49e3-8f68-6cdafc7b60c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_per_ld_block = []\n",
    "\n",
    "for _, ld_block_row in ld_blocks.iterrows():\n",
    "    annotation_n = annotations[(annotations['end'] < ld_block_row['end']) & (annotations['start'] > ld_block_row['start'])]\n",
    "    \n",
    "    genes_per_ld_block.append(annotation_n['gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e8a2e385-452f-4250-bb5c-cc3e1c542753",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#GENERATE NETWORKS\n",
    "\n",
    "#number of networks (output: num_networks)\n",
    "num_networks = 5\n",
    "\n",
    "#create networks (output: networks)\n",
    "networks = {}\n",
    "for i in range(num_networks):\n",
    "    network_n = random.sample(genes_list, random.randint(2,10))\n",
    "    prob_network_causal = random.uniform(0.5, 1)\n",
    "    networks[i] = (network_n, prob_network_causal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4176e90c-a42b-4971-8a1f-00de27d51b9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_probability_causal = [0.5]*len(genes_list)\n",
    "\n",
    "for i in range(num_networks):\n",
    "    (network_n, prob_network_causal) = networks[i]\n",
    "    network_gene_indices = [a for a, b in enumerate(genes_list) if b in network_n]\n",
    "    \n",
    "    for gene_index in network_gene_indices:\n",
    "        genes_probability_causal[gene_index] = max(genes_probability_causal[gene_index], prob_network_causal)\n",
    "        \n",
    "        \n",
    "    \n",
    "#normalize list to sum to 1\n",
    "genes_probability_causal = [float(i)/sum(genes_probability_causal) for i in genes_probability_causal]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7d19727d-e6ac-43c0-af00-805610c74fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#GENERATE CAUSAL GENES\n",
    "causal_genes = []\n",
    "\n",
    "ld_block_occupancy = [list(filter(lambda x: x in causal_genes, sublist)) for sublist in genes_per_ld_block]\n",
    "\n",
    "\n",
    "#if there is any ld block without a causal gene\n",
    "while not all(x != [] for x in ld_block_occupancy):\n",
    "    #names of causal genes (output: causal_genes)\n",
    "    causal_gene_n = np.random.choice(genes_list, 1, p = genes_probability_causal)\n",
    "    \n",
    "    causal_genes.extend(causal_gene_n)\n",
    "    \n",
    "    ld_block_occupancy = [list(filter(lambda x: x in causal_genes, sublist)) for sublist in genes_per_ld_block]\n",
    "\n",
    "    \n",
    "    \n",
    "for sublist in ld_block_occupancy:\n",
    "    random.shuffle(sublist)\n",
    "    \n",
    "ld_block_occupancy = [sublist[0:2] for sublist in ld_block_occupancy]\n",
    "\n",
    "causal_genes = list(set([item for sublist in ld_block_occupancy for item in sublist]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "586b2b71-e7ee-4234-afee-7eebe48f8da3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ef9306c9-577c-4298-8679-142def5e3aa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "   \n",
    "#GENERATE CAUSAL SNPS\n",
    "\n",
    "#generate number of causal snps and effect sizes for each snp (output: num_causal_snps)\n",
    "num_causal_snps = {}\n",
    "for causal_gene in causal_genes:\n",
    "    num_snps = random.randint(1, 2)\n",
    "    odds_ratios = [random.uniform(1.5, 5) for i in range(num_snps)]\n",
    "    num_causal_snps[causal_gene] = (num_snps, odds_ratios)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "356b1dda-f07d-48e4-9692-de7a42420032",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#GET GENOME OF EFFECT SIZES (output: genome_effect_sizes)\n",
    "\n",
    "#genome start and stop sizes based on first and last annotations\n",
    "genome_start = int(ld_blocks.loc[0, 'start'])\n",
    "genome_end = int(ld_blocks.loc[len(ld_blocks)-1, 'end'])\n",
    "\n",
    "genome_effect_sizes = np.zeros(genome_end - genome_start)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "4992c118-e010-45f3-8a5f-94d0711f70bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "for causal_gene in num_causal_snps:\n",
    "    (num_causal_loci, odds_ratios) = num_causal_snps[causal_gene]\n",
    "    \n",
    "    #annotation with only this gene\n",
    "    annotation_causal_gene_n = annotations[annotations['gene_name'] == causal_gene].reset_index(drop=True)\n",
    "\n",
    "    #bps occupied by the gene\n",
    "    gene_bps = []\n",
    "    for _, annotation_row in annotation_causal_gene_n.iterrows():\n",
    "        region_start = int(annotation_row['start'])\n",
    "        region_end = int(annotation_row['end'])\n",
    "        region_bps = range(region_start, region_end)\n",
    "        gene_bps.extend(region_bps)\n",
    "        \n",
    "    loci_bps = np.random.choice(gene_bps, num_causal_loci)\n",
    "    \n",
    "    for i in range(num_causal_loci):\n",
    "        genome_effect_sizes[loci_bps[i]] = odds_ratios[i]\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "acc03e57-1505-4511-99ae-0af57045370e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWoAAAD4CAYAAADFAawfAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAQvUlEQVR4nO3dbYxc5XnG8evCL7BgksV4Sew1zkKLLFm4ickKhbqKWieNgSDiuv0AKi1VU/lDW4k0LSkWUqu0qgylitKqUamV0EYNhaRg3BRIIQoghNRC1ziAwXZDErvxQuql1Ly0VgPO3Q9zFtbLzM7s7jkz9575/6TVzp6ZOXN7mb145j7Pc44jQgCAvE7pdQEAgJkR1ACQHEENAMkR1ACQHEENAMktrmKnK1asiJGRkSp2DQC1tGfPnpciYqjZfZUE9cjIiMbGxqrYNQDUku3Dre6j9QEAyRHUAJAcQQ0AyXXUo7Z9SNJrkk5IejMiRqssCgDwttkcTPy5iHipskoAAE1VMusDQHV27x3XLQ8c1AvHjmvV4ICu37xWWzYM97osVKjTHnVIetD2Htvbmj3A9jbbY7bHJiYmyqsQwFt27x3X9l3PaPzYcYWk8WPHtX3XM9q9d7zXpaFCnQb1xoi4SNJlkn7L9oenPyAidkbEaESMDg01nbMNYJ5ueeCgjr9x4qRtx984oVseONijitANHbU+IuKF4vtR2/dIuljSo1UWBuCdXjh2fFbb64BWTwcjattn2D5z8rakj0naV3VhAN5p1eDArLYvdLR6GjppfbxH0mO2n5L0hKT7IuKfqy0LQDPXb16rgSWLTto2sGSRrt+8tkcVVYtWT0Pb1kdEfE/S+7tQC4A2Jj/y90sroB9bPc0wPQ9YYLZsGK5tME+3anBA401Cua6tnlZYQg4grX5r9bTCiBpAWv3W6mmFoAaQWj+1elqh9QEAyRHUAJAcQQ0AydGjBlAJln6Xh6AGULrJpd+Tqwonl35LIqzngNYHgNKx9LtcBDWA0rH0u1wENYDS9dtZ/qrWN0G9e++4Nt70kM674T5tvOmhvjtNItBNLP0uV18cTOTABlCdmWZ3MOujHH0R1DMd2OCNA8xdu0EQf1/l6IvWBwc2gGowu6M7+mJEXbdz2rKQAFkwCOqOvhhR1+nABteQQybM7uiOvgjqLRuGtWPreg0PDsiShgcHtGPr+gU5CuWjJjKp0yAos75ofUj1ObDBR01kwuyO7uiboK6LuvXbsfDVZRCUWV+0PuqEj5pA/2FEvcDU+aMms1mA5gjqBaiOHzVZPQq0RusDKTCbBWiNoEYKzGYBWqP10QF6p9VjNgvQGiPqNlgJ2B3MZgFaI6jboHfaHXVaPQqUjdZHG/ROu6eOs1mAMjCiboOTzgDotY6D2vYi23tt31tlQdnQOwXQa7NpfVwnab+kd1VUS0p1XgkIYGHoKKhtr5b0cUl/IunTlVaUEL1TAL3Uaevj85I+I+nHrR5ge5vtMdtjExMTZdQGAFAHQW37CklHI2LPTI+LiJ0RMRoRo0NDQ6UVCAD9rpMR9UZJV9o+JOlOSZtsf6XSqgAAb2kb1BGxPSJWR8SIpKskPRQR11ReGQBAEvOoASC9Wa1MjIhHJD1SSSUAgKYYUQNAcgQ1ACRHUANAcgQ1ACTHaU6BinBlIJSFoAYqwFXVUSaCegpGQCjLTFcG4j2F2SKoC4yAUKZmF+qdaTswEw4mFrg2Isq0yJ7VdmAmBHWBayOiTCciZrUdmAlBXeDaiCjTcIv3jdVoswGzQVAXuDYiynT95rVq1uQIiXYaZo2gLmzZMKwdW9dreHBAVmNEtGPreg4kYk62bBhWqyYH7TTMFrM+puDaiCjT8OBA01ketNMwW4yogYrQTkNZGFEDFZn8dMYiKswXQQ1UiHYaykDrAwCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSYx71FFzhBUBGBHWBK7wAyIrWR4ErvADIiqAucIUXAFkR1AWu8AIgK4K6wCkpAWTFwcQCp6QEkFXboLZ9mqRHJZ1aPP6uiPjDqgvrBU5JCSCjTkbU/ydpU0S8bnuJpMdsfyMi/rXi2gAA6iCoIyIkvV78uKT4anXdTgBAyTrqUdteJGmPpJ+U9IWIeLzJY7ZJ2iZJa9asKbNGLBCs7ASq0VFQR8QJSR+wPSjpHtsXRsS+aY/ZKWmnJI2OjjLinqIfAoyVnUB1ZjU9LyKOSXpE0qVVFFNHkwE2fuy4Qm8H2O69470urVSs7ASq0zaobQ8VI2nZHpD0UUkHKq6rNvolwFjZCVSnk9bHSklfLvrUp0j6WkTcW21Z9dEvAbZqcEDjTf5NZa3s7If2EdBK2xF1RDwdERsi4qci4sKI+KNuFFYX/bI0vcqVnf3SPgJaYQl5xfplafqWDcPasXW9hgcHZEnDgwPasXV9KaPefmkfAa2whLxi/bQ0vaqVnf3SPgJaIai7gKXp81N1/xvIjtYH0uuX9hHQCiNqpDR9lscvfnBYDx+YqH37CGiGoEY6zVY53r1nvLSDk8BCQ+sD6TDLAzgZQY10mOUBnIygRjr9skgI6BRBjXSY5QGcjIOJSKefFgkBnSCokRKLhDBbdT5xF0ENYMGr+4Ur6FEDWPDqPqWToAaw4NV9SidBDWDBq/uUToIawIJX9ymdfXMwsc5HhIF+V/cpnX0R1HU/Igyg3lM6+6L1UfcjwgDqrS+Cuu5HhAHUW18Edd2PCAOot74I6rofEQZQb31xMLHuR4QB1FtfBLVU7yPCAOqtL1ofALCQEdQAkBxBDQDJEdQAkFzfHEwE0BucZ2f+CGoAleE8O+Vo2/qwfa7th23vt/2s7eu6URiAhY/z7JSjkxH1m5J+NyKetH2mpD22vxkRz1VcG4AFjvPslKPtiDoiXoyIJ4vbr0naL4nPLADa4jw75ZjVrA/bI5I2SHq8yX3bbI/ZHpuYmCipPAALGefZKUfHQW17maS7JX0qIl6dfn9E7IyI0YgYHRoaKrNGAAvUlg3D2rF1vYYHB2RJw4MD2rF1PQcSZ6mjWR+2l6gR0rdHxK5qSwJQJ5xnZ/46mfVhSV+StD8iPld9SQCAqTppfWyU9CuSNtn+dvF1ecV1AQAKbVsfEfGYJHehFgBAE5zrAwCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSI6gBIDmCGgCSaxvUtm+zfdT2vm4UBAA4WScj6r+VdGnFdQAAWmgb1BHxqKSXu1ALAKCJ0nrUtrfZHrM9NjExUdZuAaDvlRbUEbEzIkYjYnRoaKis3QJA32PWBwAkR1ADQHKdTM+7Q9K/SFpr+4jtT1ZfFgBg0uJ2D4iIq7tRCACgOVofAJAcQQ0AyRHUAJAcQQ0AyRHUAJAcQQ0AyRHUAJAcQQ0AyRHUAJAcQQ0AyRHUAJAcQQ0AyRHUAJBc27PnddNn/+lZPffCq70u4x2ufeVWrTvlsEbOPqPXpehm/7cOLF0iLT9/Ts/ftOuQRo6G1py5puTKOvPs0ov12rLVWrqm+td/6fD9ihMTevc5A5W/1nyd96N1Glx6jgbfu6qrr/vYK0/pv055XYvPPq2U/Q2e9aCWLXtZy5efXcr+dh7fpMMa0cDp7ytlf1W7cNmA/viC1aXvlxE1ACTniCh9p6OjozE2Nlb6fgGgrmzviYjRZvcxogaA5AhqAEiOoAaA5AhqAEiOoAaA5AhqAEiOoAaA5AhqAEiOoAaA5AhqAEiOoAaA5AhqAEiOoAaA5AhqAEiOoAaA5Dq6wovtSyX9uaRFkr4YETeVXcjIDfeVvUsA6JlDN328tH21HVHbXiTpC5Iuk7RO0tW215VWgQhpAPVTZq510vq4WNLzEfG9iPiRpDslfaK0CgAAM+okqIcl/WDKz0eKbSexvc32mO2xiYmJsuoDgL7XSVC7ybZ3XGgxInZGxGhEjA4NDc2/MgCApM6C+oikc6f8vFrSC9WUAwCYrpOg/jdJF9g+z/ZSSVdJ+nqZRZR5dBQAMigz19pOz4uIN23/tqQH1Jied1tEPFtaBQXCGgCa62gedUTcL+n+imsBADTBykQASI6gBoDkCGoASI6gBoDkHPGOtSvz36k9IenwHJ++QtJLJZZTBWqcv+z1SdRYFmrszPsioulqwUqCej5sj0XEaK/rmAk1zl/2+iRqLAs1zh+tDwBIjqAGgOQyBvXOXhfQAWqcv+z1SdRYFmqcp3Q9agDAyTKOqAEAUxDUAJBcmqC2fantg7aft31Dl1/7XNsP295v+1nb1xXbl9v+pu3vFN/PmvKc7UWtB21vnrL9g7afKe77C9vNLrww1zoX2d5r+96k9Q3avsv2geJ3eUnCGn+n+G+8z/Ydtk/rdY22b7N91Pa+KdtKq8n2qba/Wmx/3PZISTXeUvy3ftr2PbYHs9U45b7fsx22V/SyxjmLiJ5/qXH61O9KOl/SUklPSVrXxddfKemi4vaZkv5djQv5/qmkG4rtN0i6ubi9rqjxVEnnFbUvKu57QtIlalwZ5xuSLiuxzk9L+ntJ9xY/Z6vvy5J+o7i9VNJgphrVuITc9yUNFD9/TdKv9bpGSR+WdJGkfVO2lVaTpN+UdGtx+ypJXy2pxo9JWlzcvjljjcX2c9U4TfNhSSt6WeOc3yPdeqE2v+BLJD0w5eftkrb3sJ5/lPTzkg5KWllsWynpYLP6ijfBJcVjDkzZfrWkvy6pptWSviVpk94O6kz1vUuNEPS07ZlqnLz+53I1TvF7bxE2Pa9R0ohODsHSapp8THF7sRor8DzfGqfd9wuSbs9Yo6S7JL1f0iG9HdQ9q3EuX1laHx1dQLcbio8zGyQ9Luk9EfGiJBXfzyke1qre4eL29O1l+Lykz0j68ZRtmeo7X9KEpL8p2jNftH1GphojYlzSn0n6D0kvSnolIh7MVOMUZdb01nMi4k1Jr0g6u+R6f12N0WeqGm1fKWk8Ip6adleaGjuRJag7uoBu5UXYyyTdLelTEfHqTA9tsi1m2D7fuq6QdDQi9nT6lBZ1VPl7XqzGx86/iogNkv5HjY/srXS9xqLP+wk1PuquknSG7WtmekqLWnr5fp1LTZXWa/tGSW9Kur3N63W1RtunS7pR0h80u7vF6/Xs9ziTLEHd8wvo2l6iRkjfHhG7is3/aXtlcf9KSUeL7a3qPVLcnr59vjZKutL2IUl3Stpk+yuJ6pt8zSMR8Xjx811qBHemGj8q6fsRMRERb0jaJemnk9U4qcya3nqO7cWS3i3p5TKKtH2tpCsk/XIUPYFENf6EGv9Tfqr421kt6Unb701UY0eyBHXlF9CdSXFU90uS9kfE56bc9XVJ1xa3r1Wjdz25/ariKPB5ki6Q9ETxEfU12x8q9vmrU54zZxGxPSJWR8SIGr+bhyLimiz1FTX+UNIPbK8tNn1E0nOZalSj5fEh26cX+/6IpP3JapxUZk1T9/VLarx/yvikd6mk35d0ZUT877Tae15jRDwTEedExEjxt3NEjUkDP8xS42z+MSm+JF2uxmyL70q6scuv/TNqfIR5WtK3i6/L1eg/fUvSd4rvy6c858ai1oOacsRf0qikfcV9f6mSDzZI+lm9fTAxVX2SPiBprPg97pZ0VsIaPyvpQLH/v1PjqH9Pa5R0hxo98zfUCJNPllmTpNMk/YOk59WY0XB+STU+r0bPdvJv5tZsNU67/5CKg4m9qnGuXywhB4DksrQ+AAAtENQAkBxBDQDJEdQAkBxBDQDJEdQAkBxBDQDJ/T+2EJ0VqGpiQgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plot -log10 p values\n",
    "plt.scatter(range(len(genome_effect_sizes)), genome_effect_sizes)\n",
    "\n",
    "#show ld blocks as lines\n",
    "for i in range(len(ld_blocks)):\n",
    "    x_values = [ld_blocks.loc[i, 'start'], ld_blocks.loc[i, 'end']]\n",
    "    y_values = [0.5, 0.5]\n",
    "    plt.plot(x_values, y_values)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a615e869-4eaa-4818-bef5-7dd8300a99c8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "545a9663-9062-4b5e-b348-7d14fc11d6a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/7j/329_cmn56qz2jnvgvylvq3dm0000gn/T/ipykernel_24619/1138240172.py:44: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  a = np.float64((2*EVX2-EVX*EVX))/(EVX2-EVX*EVX)\n",
      "/var/folders/7j/329_cmn56qz2jnvgvylvq3dm0000gn/T/ipykernel_24619/1138240172.py:45: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  b = np.float64((EVX2*EVX))/(EVX2-EVX*EVX)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of nan z scores:  1458\n",
      "0\n",
      "number of nan z scores:  819\n",
      "1\n",
      "number of nan z scores:  9\n",
      "2\n",
      "number of nan z scores:  19\n",
      "3\n",
      "number of nan z scores:  1331\n",
      "4\n",
      "number of nan z scores:  1614\n",
      "5\n",
      "number of nan z scores:  33\n",
      "6\n",
      "number of nan z scores:  19\n",
      "7\n",
      "number of nan z scores:  23\n",
      "8\n",
      "number of nan z scores:  111\n",
      "9\n"
     ]
    }
   ],
   "source": [
    "p_values_genome = []\n",
    "\n",
    "for ld_block_index, ld_blocks_row in ld_blocks.iterrows():\n",
    "    \n",
    "    n_snps = int(ld_blocks_row[\"end\"] - ld_blocks_row[\"start\"])\n",
    "    \n",
    "    n_haplotypes = 100\n",
    "    neighborhood_length = 5\n",
    "    maf = [random.uniform(0.05, 0.50) for _ in range(n_snps+neighborhood_length)]\n",
    "    \n",
    "    \n",
    "    causal_indices = list(np.nonzero(genome_effect_sizes)[0])\n",
    "    causal_indices = list(filter(lambda x: ld_blocks_row[\"start\"] <= x <= ld_blocks_row[\"end\"], causal_indices))\n",
    "    adjusted_causal_indices = [int(x - ld_blocks_row[\"start\"]) for x in causal_indices]\n",
    "\n",
    "    causal_variants = [\"snp\" + str(int(s)) for s in adjusted_causal_indices]\n",
    "    odds_ratios = [genome_effect_sizes[i] for i in causal_indices]\n",
    "    \n",
    "    n_controls = 100\n",
    "    n_cases = 100\n",
    "\n",
    "    p_values = simulate_p_values(n_snps, n_haplotypes, neighborhood_length, maf, causal_variants, odds_ratios, n_controls, n_cases)\n",
    "    \n",
    "    p_values_genome.extend(p_values)\n",
    "    print(ld_block_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2a7049d-8e91-452c-b3b0-7afb9a24f948",
   "metadata": {},
   "outputs": [],
   "source": [
    "# #===================\n",
    "# #JUST ONE LD BLOCK\n",
    "# #===================\n",
    "# p_values_genome = []\n",
    "\n",
    "# ld_blocks_row = ld_blocks.iloc[2,:]\n",
    "\n",
    "# n_snps = int(ld_blocks_row[\"end\"] - ld_blocks_row[\"start\"])\n",
    "\n",
    "# n_haplotypes = 100\n",
    "# neighborhood_length = 5\n",
    "# maf = [random.uniform(0.05, 0.50) for _ in range(n_snps+neighborhood_length)]\n",
    "\n",
    "\n",
    "# causal_indices = list(np.nonzero(genome_effect_sizes)[0])\n",
    "# causal_indices = list(filter(lambda x: ld_blocks_row[\"start\"] <= x <= ld_blocks_row[\"end\"], causal_indices))\n",
    "# adjusted_causal_indices = [int(x - ld_blocks_row[\"start\"]) for x in causal_indices]\n",
    "\n",
    "# causal_variants = [\"snp\" + str(int(s)) for s in adjusted_causal_indices]\n",
    "# odds_ratios = [genome_effect_sizes[i] for i in causal_indices]\n",
    "\n",
    "# n_controls = 100\n",
    "# n_cases = 100\n",
    "\n",
    "# p_values = simulate_p_values(n_snps, n_haplotypes, neighborhood_length, maf, causal_variants, odds_ratios, n_controls, n_cases)\n",
    "\n",
    "# p_values_genome.extend(p_values)\n",
    "# print(ld_block_index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c83ee3e-41e5-487e-a313-e04a445ca4ea",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "ab947bd5-96d5-48d6-a78a-993de4845378",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "causal_variants:  ['snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374', 'snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374', 'snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374', 'snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374', 'snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374', 'snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374', 'snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374', 'snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374', 'snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374', 'snp1187', 'snp1188', 'snp1595', 'snp1602', 'snp1897', 'snp2667', 'snp3681', 'snp5543', 'snp6326', 'snp6438', 'snp6629', 'snp8855', 'snp8856', 'snp8880', 'snp9953', 'snp11703', 'snp11974', 'snp12080', 'snp12890', 'snp13374']\n"
     ]
    }
   ],
   "source": [
    "# ld_blocks_row = ld_blocks.loc[0,]\n",
    "# causal_indices = list(np.nonzero(genome_effect_sizes)[0])\n",
    "# adjusted_causal_indices = [int(x - ld_blocks_row[\"start\"]) for x in causal_indices]\n",
    "# causal_variants = [\"snp\" + str(int(s)) for s in adjusted_causal_indices]\n",
    "\n",
    "# #only 2 cvs for this ld block\n",
    "# causal_variants = causal_variants[0:2]\n",
    "\n",
    "# print('causal_variants: ', causal_variants)\n",
    "\n",
    "causal_variants = []\n",
    "for i in range(10):\n",
    "    ld_blocks_row = ld_blocks.loc[i,]\n",
    "    causal_indices = list(np.nonzero(genome_effect_sizes)[0])\n",
    "    adjusted_causal_indices = [int(x - ld_blocks.loc[0, \"start\"]) for x in causal_indices]\n",
    "    causal_variants_n = [\"snp\" + str(int(s)) for s in adjusted_causal_indices]\n",
    "    causal_variants.extend(causal_variants_n)\n",
    "#only 2 cvs for this ld block\n",
    "causal_variants\n",
    "\n",
    "print('causal_variants: ', causal_variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9995f5d2-2276-4194-b636-a489367f4d67",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "4fd5d3d5-1246-429b-be0b-39154960f64d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/7j/329_cmn56qz2jnvgvylvq3dm0000gn/T/ipykernel_24619/4106349312.py:5: RuntimeWarning: divide by zero encountered in log10\n",
      "  neg_log_p_values = -1*np.log10(p_values_genome)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAp10lEQVR4nO3de3xU9Z3/8deHhElIwiWBcBECwYjS0CpqtF66hSptRW3dXrD2125Z2131t73+tK1S7WV3tWpda93ur1Zsq3S3q4Xarf4sVi3epYKRAmIkQiAQBCGQcMkFJpfv749zZpiESSYkM5nh5P18PPKYM99z+8wk+cx3vud7vl9zziEiIsEyLN0BiIhI8im5i4gEkJK7iEgAKbmLiASQkruISABlpzsAgHHjxrnS0tJ0hxHX6697j2effXzrE+030PNmyjGTIVnvVUTkOJn6ehNJVdyZ9n7Eiye2bLD+hzLtfTker7/++l7nXHG8dZaoK6SZlQC/BiYCncBi59y9ZvYD4B+Ben/T7zjnlvv7LAK+BHQAX3POPdXbOSoqKlxlZWXfX9EgMvMee3qbelqfaL+BnjdTjpkMyXqvIiLHydTXm0iq4s609yNePLFlg/U/lGnvy/Ews9edcxXx1vWl5t4O3OCcW2NmI4HXzewZf909zrl/63aycuAqYBZwEvBnMzvVOdfR/5cgIiLHI2Gbu3Nul3Nujb98CHgLmNzLLlcAjzjnjjjntgKbgXOTEayIiPTNcV1QNbNS4ExglV/0FTNbb2a/MrNCv2wyUBez2w7ifBiY2TVmVmlmlfX19d1Xi4jIAPQ5uZtZAfAo8A3n3EHgPqAMmA3sAu6ObBpn92Nas5xzi51zFc65iuLiuNcDRESkn/qU3M1sOF5i/41z7vcAzrndzrkO51wn8ABHm152ACUxu08BdiYvZBERSSRhcjczA34JvOWc+3FM+aSYzT4BbPCXHweuMrMcM5sOzABWJy9kERFJpC+9ZS4E/g54w8zW+mXfAT5rZrPxmlxqgWsBnHNvmtlSoAqvp82X1VNGRGRwJUzuzrmXid+OvryXfW4DbhtAXCIiKdHQHGZZZR0LKkooyg/1e5tMp+EHRGRIWVZZx+1PbmRZZd2Atsl0GTH8gIjIUY77X9gCnEz8RoOea9YNzWGWrNzq7zcj7v4LKkq6PMbTl20ynZK7iGSc25/ciJfc44vUrAGunVPWpfzeFZv9ZzPi7luUH+qyT3+3yXRK7iKScRbNn8l1d/a8vqea9YKKElrC7YBxfS/7DwUJBw4bDBo47PjPmynHTAYNHNaVBg7TwGF91dvAYbqgKiISQEruIiIBpOQuIhJASu4iIgGk5C4iEkBK7iIiAaTkLiISQEruIiIBpOQuIhJASu4iIgGk5C4iEkBK7iIiAaTkLiISQEruIiIBpOQuIhJASu4iIgGk5C4iEkBK7iIiAaTkLiISQEruIiIBpOQuIhJASu4iIgGk5C4iEkBK7iIiAaTkLiISQEruIiIBpOQuIhJACZO7mZWY2XNm9paZvWlmX/fLi8zsGTPb5D8WxuyzyMw2m1m1mX00lS9ARESO1Zeaeztwg3PuPcB5wJfNrBy4CVjhnJsBrPCf46+7CpgFXAL8zMyyUhG8iIjElzC5O+d2OefW+MuHgLeAycAVwBJ/syXA3/rLVwCPOOeOOOe2ApuBc5Mct4iI9OK42tzNrBQ4E1gFTHDO7QLvAwAY7282GaiL2W2HX9b9WNeYWaWZVdbX1/cjdBER6Umfk7uZFQCPAt9wzh3sbdM4Ze6YAucWO+cqnHMVxcXFfQ1DRET6oE/J3cyG4yX23zjnfu8X7zazSf76ScAev3wHUBKz+xRgZ3LCFRGRvuhLbxkDfgm85Zz7ccyqx4GF/vJC4LGY8qvMLMfMpgMzgNXJC1lERBLJ7sM2FwJ/B7xhZmv9su8AdwBLzexLwHZgAYBz7k0zWwpU4fW0+bJzriPZgYuISM8SJnfn3MvEb0cHuLiHfW4DbhtAXCIiMgC6Q1VEJICU3EVEAkjJXUQkgJTcRUQCSMldRCSAlNxFRAJIyV1EJICU3EVEAkjJXUQkgJTcRUQCSMldRCSAlNxFRAJIyV1EJICU3EVEAkjJXUQkgJTcRUQCSMldRCSAlNxFRAJIyV1EJICU3EUkIzQ0hwEHOH8ZwDH3ruf43h820NAcpqE5zP0v1FBT3xT38eh+knCCbBGRwbCssg44OWa5DIDafS3U7tvG5MIRANz+5EZe3bKP56rrj3kEuHZOWTrCzzhK7iKSERZUlHBdnOXSsXl8cEYxCypKotvOK5/AeSfvPuYxdpuhzpxz6Y6BiooKV1lZme4w4jLzHnt6m3pan2i/gZ43U46ZDMl6ryIix8nU15tIquLOtPcjXjyxZYP1P5Rp78vxMLPXnXMV8dapzV1EJICU3EVEAkjJXUQkgJTcRUQCSMldRCSAlNxFRAJIyV1EJICU3EVEAkjJXUQkgBImdzP7lZntMbMNMWU/MLN3zGyt/3NpzLpFZrbZzKrN7KOpClxERHrWl5r7Q8Alccrvcc7N9n+WA5hZOXAVMMvf52dmlpWsYEVEpG8SJnfn3ItAQx+PdwXwiHPuiHNuK7AZOHcA8YmISD8MpM39K2a23m+2KfTLJgN1Mdvs8MuOYWbXmFmlmVXW19cPIAwREemuv8n9PrzBlmcDu4C7/XKLs23csdacc4udcxXOuYri4uJ+hiEiIvH0K7k753Y75zqcc53AAxxtetkBxA6oPAXYObAQRUTkePUruZvZpJinnwAiPWkeB64ysxwzmw7MAFYPLEQRGeoi0+s1NIepqW/i6gdXU1PflO6wMlrCmZjM7GFgLjDOzHYA3wfmmtlsvCaXWuBaAOfcm2a2FKgC2oEvO+c6UhK5iAwZyyrruP3JjQAxU+pV8eDV6q/Rk4TJ3Tn32TjFv+xl+9uA2wYSlIhIrMj0eQsqSphXPgGo4pbLy9MbVIbTHaoikkEc4GhoDncpLcoPce2cMoryQ5QVF3D3lbP5c9XuY7aTo5TcRSTjLKusS7j+9ic3JtxuKEvYLCMikmoNzWGWrNwKnArAO42tRHpRf/3hv/LuwSNMG5vHzv2Huf7Dp/JOYwulY/M4p7Qouv/Pn9/Mq1v20djSxr1XnQkUxj/ZEKHkLiJpt6yyjntXbCaS3H/96ja8UUzgsXW7AFi11btRfteBVmrqmwH46bObePDqc1lWWcfil7ZGj/et363D6wcydCm5i0jaLagooSXczvV3es+v+Zvp3OwvX3HGpGNq7n/asIuqXYeiF1UXVJSwr+lItOZ+16fP4OxvpunFZAhzLu4NpIOqoqLCVVZWpjuMuMy/57ant6mn9Yn2G+h5M+WYyZCs9yoicpxMfb2JpCruTHs/4sUTWzZY/0OZ9r4cDzN73TlXEW+dau4iklbPb9zDN5b+lfOmjwXOBvCbXfIBuP+FLdGukEtWbqU13Mn+ljB/2bKPWSeN4s2dBzn/5LFMGjOCOacW89NnN/k1+oL0vKAMoeQuIml1/bK17G9p509v7o6W3fpEFXAOQPTmJcBvlz+qrrHVe3x9BwBPrN/pfzBUMdQHpFVyF5G0+vGC2dGa+/1+2S2Xl/OQv7xo/sxozb0l3N7nmvtDcc41lKjNPQG1uaee2ty7Upu72tz7Sm3uIpLRauqb+N4fNgDv90uOHT08sk3Z+AIK80IsvKAUwO8fbyy8oJSi/NBghZzxlNxFJO1ufaKKV2r29WmbyHZ5IW8Gz0g7fF4oi2vnlKU20BOIkruIpN0tl5cTbt/Atj5sE6m5x7bDg0Wfi0dt7gmozT311Obeldrc1ebeV721uWvgMBGRAFJyFxEJICV3EZEAUnIXEQkgJXcRkQBSchcRCSAldxGRAFJyFxEJICV3EZEAUnIXEQkgjS0jIhlhzbZGYAwAn/zZSuACAD73wKroeDJzTi3m7qermVw4Ijqf6mu1Dcwrn8Dja98hWaNDNjSHWVZZx4KKkhN2pEkldxHJCN/63TpgDgBrtu+PlseOBHl0piXPrgOt1NQ38+qWfTxXXQ8kZ3TIZZV10RmgTtSRJpXcRSQj3PXpMzj7m97yWVPHREeIvLBsbJ9q7qdP8WruyRgdMnKME3mkSY0KmYBGhUw9jQrZlUaF1KiQfaVRIUVEhhgldxGRAFJyFxEJICV3EZEASpjczexXZrbHzDbElBWZ2TNmtsl/LIxZt8jMNptZtZl9NFWBi0gwNDSHueeZau555m3AAY6G5nC3ZW+7+1+ooaE53GW5+zrx9KXm/hBwSbeym4AVzrkZwAr/OWZWDlwFzPL3+ZmZZSUtWhEJnGWVddy7YjP3rtjUpaz7cqTv+bLKui7L3deJJ2E/d+fci2ZW2q34CmCuv7wEeB640S9/xDl3BNhqZpuBc4G/JCleEQmYBRUltITbAeP6O4+WXRezPt5jonVDXZ/6ufvJ/Qnn3Hv95/udc2Ni1jc65wrN7D+AV51z/+WX/xJ40jn3uzjHvAa4BmDq1Klnb9u2rfsmGUH93FNP/dy7Uj939XPvq8Hs525xyuK+Zc65xc65CudcRXFxcZLDEBEZ2vqb3Heb2SQA/3GPX74DiP1eNAXY2f/wRESkP/qb3B8HFvrLC4HHYsqvMrMcM5sOzABWDyxEERE5XgkvqJrZw3gXT8eZ2Q7g+8AdwFIz+xKwHVgA4Jx708yWAlVAO/Bl51xHimIXEZEe9KW3zGd7WHVxD9vfBtw2kKBERAbOu9zX0NyW8jHZM3H8d92hKiKBNhh93zOxn73GcxeRQBuMvu+Z2M9eyV1EAsrrmT0YzSRF+aGMm7FJzTIiIgGk5C4iGczFHRCspr6Jqx9cTU19U7QsdgAyDSCmZhkRyXDxJqq+9Ykqf0LsKh68+lzg6ABk4E2SDZnVTDLYlNxFJKMtmj/zmAuVt1xeDlT5j57YAchiBx4bqjRBdgIaOCz1NHBYVxo4TAOH9ZUmyBYRGWKU3EVEAkjJXUQkgJTcRUQCSMldRCSAlNxFRAJIyV1EJIB0E5OIZCiv4/kP/+jdoToilMXCC0opyg9l5PjpmUbJXUQy2uKXtkSX80JZXDunLDp+OpBxozFmCiV3EUmrhuYwS1ZupbG5DZgFwAW3PwtcBEBxfogO10l7J4zKzebqB1ez8PxSzi0t5Kk336Vq5wHW7TjAj6+czVnTCtP3QjKMkruIpJHjhqVr/UHAIJLcdx44HN2iPmaEx+8/toFwJ7xW20DTEW965jXb9wPwqftWUlwQYkpRHnctOAMoGIwXkLF0QVVE0uq56nouLBvL5DG50bKcrK7bDPfHfxk+3FsRSeyxHLCnKcya7fu59YmqVIV7wlByF5G0WjR/Jj/9X2fx6y+9P1p2/9+dE10+a+oYfnvdBXz94lO47L2TKCkcwZVnT+GKMyYxMjeL8kkjueKMSUwYmcP4ghBnTR3jjxbpAHdcY7s3NIe7jB/f/XmseGPK9+WYg0XJXUTSyLh2ThlF+SHKio82oyz5S210+ff/dCFnTSskL5TN0td38PnzpvGjBWdQftJoDh3uoGrXIcpPGs2qm+ex+pYP8/t/urDLsY5n0uruE133NvF1ZEz5RN8S0jV5ttrcRSTj3HJ5OQ91K+s+CXX38dt7cjyTVsc7R0/HiDemfF+OOVg0nnsCGs899TSee1caz13jufeVxnMXERlilNxFRAJIyV1EJICU3EUkoAbeFTJVBuM8Su4iEmgD6QqZKoNxHnWFFJFAG0hXyFQZjPOoK2QC6gqZeuoK2ZW6QqorZF/11hVyQDV3M6sFDgEdQLtzrsLMioDfAqVALXClc65xIOcREZHjk4w29w8552bHfHrcBKxwzs0AVvjPRURkEKXiguoVwBJ/eQnwtyk4h4hIxoj0fqmpb4rbCyYdg4cN9IKqA542Mwfc75xbDExwzu0CcM7tMrPx8XY0s2uAawCmTp06wDBERNIn0vvl1S37omPTx84QlY6Zowaa3C90zu30E/gzZraxrzv6HwSLwbugOsA4RETSJtLrZV75BM47efcxvWDSMXjYgJK7c26n/7jHzP4HOBfYbWaT/Fr7JGBPEuIUEclYRfmhaI28bM6xM0DFrh8s/W5zN7N8MxsZWQY+AmwAHgcW+pstBB4baJAiInJ8BlJznwD8j3mdRLOB/3bO/cnMXgOWmtmXgO3AgoGHeSLyWppKb1repXREFrQeO0MYI7KNs6cVsXrrPsKdAJcC8J7vPkm4zdEBnFacx6b6Fjpj9ssG2mOeZ+H1S43IzTY+PnsyN81/DxAa6IvKUF1b9Rqa2yjKP5Ffq/d6auqbu0w6MTR4r72huQ0YHl3u3+/z6LF637+v251Y+l1zd85tcc6d4f/Mcs7d5pfvc85d7Jyb4T82JC/cE1+8xA7Q2u54uSaS2GPK/cQOUN0tsUPXxA5dEzvA4XbH0sodgz4LTDoF5bUOpXlAG5rD3PNMdfR57O+wt99nX3qhLFm5tU89VZZV1sXt9RJ7jp7Ol2i/44k3WTT8wCDre83dLx9uSam5L6go4bpkvpAMNtgz3qRKohl+gsFx5c//wq4Dh6lrbAVOBeDh1duBkwF46s13+eP6nYwIZfP590/lJys28d3LyqnefYh9TWEWv7SFZzfuZuKoXF7ctJd7rpzN3JmxnfSsTz1VFlSUxO31AkT3j12O1xump/0i2w5mrxkl95Tx7mmuveOy/u19p/f41r9emqyAAqzr+AMn/ldr7/UMlSaZ1bXeDex5w482JNTua4kur9m+P7q8aqvXEPDFh16jExiTl+2XH70J/ssPv87XLjoV78PBWHhBKXmhrF4+9L33uyg/1Guvl56WY5/3db/BqIBobJkENLZM6mlsma6G1tgyjlNvfpIj7Y6cLOPtH84HYNqNy9l256XR5eMxMieLQ0c6/P1tyI4toyF/RSSNjIf/8XzKivP5twVnREuLY759FeeHGDMiiywgLzSMK86YxEWnFZNlcPFpxXzhvGmUTywgL2RMHJ3LTz97Fovmz/T3dnHbvHu6kzSR/rSZD/Sc/aVmmZTxqgF3Ln+L37y2ndKxBXz2nBLufGojc2YUUzqugIUXlNLYEuZ7f9jArMmjuW5OWbcmBcfnHlgVXbe+bj/XL1vLjxd4bYo19U187w8bMOCVmn2Myctm+rgCbrmsnBfe3kNjcxuvbtnL9oZW7vr06cDkdLwRIr06a1ohK26Y26Xste9+GPve0eXjNXfmeK7z/weXVdYd0+bdtW28723f/WkzT3T3aqoouafYfS9uBWD9jgNU7TxIe6fjsXW7AMgLZfHqln28UuP9jI1zo0PsuvtfrKGhuY3rl61lzXc/wq1PVPFKzb7oto0t7TRu38+3freOmvrmLsf59qPrUXKXoShem3ds2/jxdDToT5t5ortXU0XJPcX+9wen91hzX1BRwrzyCYTbvZp7vF/6hWVjo+tOmzAyWnMHrzdFuL1vNfcffep0Pn7r4L52kYFz3P/CFhZUlPTjQvnRC6URkTtFe2oaaWgOs6yyrsv5GprDLFm5ldZwJyNCWSy8oJSi/BA19U3c+kQVt1xeHvfi95ptjVy/dC3nlhYxacwIDrS08eLb9byzv5XCvFD0OKmiC6oJ6IJq6umCaldD64Jqosk6HNNuXM6i+TOPuymjt9d7/ws13P7kRhbNn8l1c8ui28WWR84XKYuIrLv6wdU8V13Ph04r5sGrzz3mHBff/XyXb9BlxfldnvfnNR37GlM0WYeISKotmj8z6U0Zsc0r1/VQHlvWEm6P1twj67z7EKp6vB/hrk+f0aXmPufUYu5+upqy8QUU5oVS3jyjmnsCqrmnnmruXanmPtjT7Hkb7Gtq69IU0715JhOpK6SIDFGOqx9cTU19EwA19U1c/eBqnt+4p0s5dB1+YMnKWm5/ciM3LF1LTX0T9zxTzT3PvB0dViD2eXfduz7G7h9blupukWqWEZFA87ofVvHg1edy6xNVPFddz7od+/3ByaqAc4Cuww98/eJT+NBpxdF9I10Y80JZANy7YnP0efd28+5dH2O7QK7fsf+YslR1i1RyF5G0amgOc8fyKp54Yyfg3aH6uQdWAe8H4JM/WwlARWkRl8yayN1PVzNr8mgumTWRnz67iVsuL6cwL8TPn9/Muh0HOKNkDJ85pwTIB7wLmV+9aAYNzWFmjC+grcPxDx+YzgMvbaGkKK9LLC3hdr5+8YzoPShQxcLzSwm3dzJr8mjOKS3i7qerubJiCtv2tbDPr8nHNt107/o4r3wCM8Zvp2rXIb560QxOnzKa1nAnp08Zk9J2dyV3EUmrf358Q/Tej4jY+zciY8us2b6fB17cgvPXL35xCwDPVb9AyCDst62v2trA4+t2AhcB3tDJr9U28FptA4tf2sqi+TOZO3M81bsP+b1gZgFejfveFZtZNH8mRfkhllXWRWvXr9Ts44OnFvPTZzfxSs2+aM+XVVsbjrk/Jd7EHWMLcnh581b+ZsY48kLZ3LtiY/Q8qaLkLiJp9XTV7j5v29O10XC3Fe8eOBxd7t7bJrZHTEu4nevvPLY89jH25qN55ROAKr560QxeeHsPYH2qfffUCyeVdEFVRNLqR586ndAwGNdDLXZUbhbDh0FJ4QhG5mbx/tJCsmLWDwMKcoaRbd62BTlZXFkxJbq+px4vjS1h1u844D+zaI27KD9EQ3OYnz+/mZc27eVASxsAtXubozctjc4bTmVtI40tYZasrI1eGO1p7JnIsRtbwtywdC3zyiekvBeOau4iklYfmz2Zj832hsaIjCdTe8dl0WGv1//gkn4d965uY8t0HxcmcnE1nmWVdSx+yRs6ZNeBVn9WrMhNSN4kKpGhQeDohdVEY88cPWdV3BufkknJXUQCLXLRc175BF7dso9zSou4/4UavnrRDAAeAsBRU9/Mn6u85pcFFSXsazrCuh0HOG3iSC6eOYHDbR1MGj2CWy4vp25fC69vb2DiqFwK83L85pquzS+RvvLzyifw56rIhVXvgu4tl5envC+9kruIBNriF7cw1k+ekZr6c9X1LJo/kwevPpeHvujV8GNr8tfOKeM7l5VHhx6IdItcNH8mZcUF3PpEFQdbOzjY2gx4Hwplcwq6XEyN7Nu9S2TkGLHDGqSiO6SSu4gESqRGHJmm76ypYzintIg/bdjFB04ZF+2OuK8pzA//+Bbgjf3+1YtmMGN8AS9t2su88gmUFRdEa/CV2xopn1jAL17cwq//UstNl8xkf0uY2r3NtLR1cqStnYvvfj46/d+CihLOKS2irDifj5RPYHtDC588czLNR9p5duMe5pVPiF7QbQl3HNOdMhmU3EUkjRzz73mBTbubyc4G8GZfmn7T8ujyzJuXc6TD6ymTn2O0HnE4IHe4UZgXYueBI4B3YTUnG1qjkwp7yX3N9v389NlN0Vp5bHdEj5fcX6ttYNOeJl7evJdFj64nPyebWy4vZ9Oepi5T/QH8ZMUmxhWEaPRP9pMVm+l0cP2ytf7NUfDqln3U1Ddz19PVNDS38e1H19Pa5s2AfM2vK1l23QXkhbK5/cmNcW+GGigldxFJq7d2eyMltsfM9B7bs/FwzIzvzUeOrmltc7T6iR2gk9jEftQ1HzyZz5xTwulT3iHSdbGxJcyLb9dTNr6Af/W3m1c+gXcaWygdm8fh9k5W19bTEn6DM6aMpiXcwcRROby6pYHsLOOuT5/BnX/yPhyyhxk3XXIaD79Wxz98YDq/eHkrp00Yyb6mI9H9nq7aTWtbJ6FhkDM8i5r6ZpasrGXhBaVAarpFKrmnjPdHeOa/PMO/fHwWr9U28Oe39jCuIMR5ZeN430mjuPmxN5gyJo+K0iJys4927trfEgZOB7y78ypKi/jMOSXRizL3P7+Zp6p2876TRvPmroN8eW4Zv3qllvaOTiaOzmX21EIOhzt4efNexozIZuSIED/85PuAoTHh8onP+9tpaG7L6EGrkuU9E/Jjau6e2LHgcrM47pr75NG5bPP3X7u9kVdr9pKdNYyK0iIAHl+7k1dq9rF1bzORm5i+tWxdtIY+Nm84AHubDrP4pQa+cN40CvOHk5Odxcub9/KDxzcwc9Io2jo6GZ41jPycbBpbwvznq9uoqW/ma4/8lYOH2ykdmxed1Hv4MAh3QviI92m1rLKO1raOODOwJYdGhUyg/6NCegXTblzOiOHDol/HIuKVxeo+OXDkgs7R8S6OGmbQmeDX+KHTinnoi+f2+lrSRaNCdhX52/n581uS+lU9096PVI0Kec8zb3Pvik09TrA9Ji+bk0blUvWuN2hYbxNxR/63Jo7K4d2DR45ZP60oj20NLWQPM9o7HVkGHd3iLSkcQV1ja4/xfui0Yu6+cna/ErzGc0+jwrzh/aq53+UvnzV1TLTmHhmnorjg+Gvut1xe7nf5khNF8r+qD41vBK3hrm0zo0ZkkTc8m3cPHiF7mLG/pZ39LU3H7JcfGkZzuJP3Ty9k274W3j14JFpp2tvkJfYsICc0jCmFI8g2Y1tDCwCnTx7FmroDdDjIzR7G4Xav4mYx+8Yqn1jA/tZ2zLyeO7HzvCaLau4JaDz31FPNvavUjeeemm8E/RWZaenK+1ayqnY/cLQWXXbjcmr85ek3Lif2O+74/BCNLWFyQsM40tZJQU42ja3tDIMu20WOtXlPM4V5IW5YupbnqusZk5fN1MI81r9zkNEjsln/g48A8In/+wofnTUp2ib/qfteYX9Le7RWHqv7N+jCvOE0+neygldb39t0pMu381CWEe5wlBSOYP77JkXHxlHNXZJgaNTcpHeDNUFzYt7fYySxx4qte3dvvNzj39rfdsRbE+mx0lMj529X1wEwKjebkblZnDR6BDMnjWJEKJtpY/NY72939QXTea22gfk/eZHsLOMrc0/h35/bxOjc4exqPEx2NuQOz6ato5PtDc2MzRvO4fYOysaPjM6PfFZJIVv2NvODj81i+Rs7+eOGXWSZcbitk9Jx+by9u4m6xlbW1TXyhfOmsWHnAVrCHTS2qCukJEEqvgLKieDYCaMzwZTROew40LXpIjYxda+Rg/dK8nN6r7lHLH5pS5fnVbsOUbXrEIB/sdPrvPCjp6q7tI3f/uRGOhwcbPUugLa3Q1tHOx0Oaupbotut33GAPQcPe809rWFq97Xw1UfWcMjv5mN4H2Nv725i9IhsDrS2s2prIxt2HqDZ/4C69YnkD0eggcOGoMypuYnApyqmsuKGOXzhvKnRsqyso21tX734FK48ewonjc7l1PH5TBiZw/smj+LimRMoHplLSVEel8yawOi84Vw/75SE5xveQ3n3i57dL4z2VAZEL7au2e4NRHYopv9m7C4HYvpqRhJ71jBj4fmlvcbcH6q5DymZWXNTc9FQ5dVp712xiV+8XOMnu/cCcCQmi0ZmPYq1+9AR1r9zEICdBw5Hl3/852O37a4t4RaDq6PT8e/PbmLuzPFJPe4Jn9zn3rmC2sbDiTc8DpGvUR7vokzpTcd2k+p9vVe+Ztt+zppWmNT4gkrNRUNXpBY7VNUfOrZHzUCd8Mk92Ykdep4QoD8+ed/Kfu6Z6EPlOI82azzgXVT3hi/NvBua1Fw0MKkeZTBV4rW5dzcMCGV1vVu1ayXsxPavV7w36cdMWXI3s0uAe/G6hv7COXdHKs5TWpib4pp7MCx/c090+e9/tZrHvvKBaAKoqW9i0aPraTrcxqEjHdx71ZmcNa2QhuYwS1ZuBYyFF5QOOGH0nHwyp7nozuVV3Pfi1jRH4X2wn/ad5XSv0OZ2S3A5WfCZc6ZxTmkh31y2liMd8NSbu3jvSWNYW9fIrv2HycoyPlI+kcidmA+v2s4vXt7KNy6eweIXa9hcfwizYZSOzWfG+AJeq21k4uhcvnbRDB54aQuTC0eweY/XL/y9k0dTmBeK/j38v7Xv8O1H1/OjT53Ox2ZPpqa+KTqhRWFeKPr7Bliyciu79h9mdW0D137wZMBrY0+U2MG7UBr7uiFY/6PLN+xKerNMSvq5m1kW8DbwYWAH8BrwWedcVbztg9nP/egdqv3R211z/ZWKYyZDX+K6qmIyd3x6dtx1yeznXnrTH/u+cYqk6veUab//ePHElg003r7unwnvS35oGG/+y/zj3i8d/dzPBTY757b4ATwCXEFkCpMhwcsstXdc1r+9/Vlo+rt/ROyY0SeyRyrf6TG5J9P//uD0DKi5e3KGkbDmLkFhiTc5TqlK7pOBupjnO4D3x25gZtcA1wBMnToVSY0FFSWBSO6D5cZLy7nx0vK0xhD5YK/+Ye8f7JnwLSOZ8kNHe2aPzRseHfirrDifLPPmMA37H3Yjso0Jo3PZfbAVM284gNzsLHYfPMLuFFycTLXL3jcx6cdMVXKP9zHU5cuxc24xsBi8ZpkUxTHkFeWHKMwdRuPhE7s3QsnonHSHkHFOK86jOuZmmhPZ+aVjePi6CzF//N3Xv/cR7Pve8oob5vb5OJGJrZ+v3hP9cMgLGR0djjZ/ZMlsgzYHhXnZFBfkRLc7dXw+dY3NdHYCNowvXjCNJa9u5UjYG0/mvOljWVe3nwOtbeSGhnE43EkoGyKjDnT/VhUysGHQ0Ul0+zF5IS48ZSwb3jlAzV7vd/c3p4zlphRUKFLV5n4+8APn3Ef954sAnHO3x9s+mG3uve830PNmyjGTQWPLdJW6sWVSc9z+StWokL0dfyDbZaLe2txTdYfqa8AMM5tuZiHgKuDxFJ1LRES6SUmzjHOu3cy+AjyF1xXyV865N1NxLhEROVbK+rk755YDmdHnSkRkiNHAYSIiAaTkLiISQEruIiIBpOQuIhJAGTGHqpnVQ/ReguM1DtibxHBSQTEmh2JMDsWYHJkQ4zTnXHG8FRmR3AfCzCp76sSfKRRjcijG5FCMyZHpMapZRkQkgJTcRUQCKAjJfXG6A+gDxZgcijE5FGNyZHSMJ3ybu4iIHCsINXcREelGyV1EJIBO6ORuZpeYWbWZbTazmwbxvCVm9pyZvWVmb5rZ1/3yIjN7xsw2+Y+FMfss8uOsNrOPxpSfbWZv+Ov+3az76OQDjjXLzP5qZk9kYoxmNsbMfmdmG/338/wMjPH/+L/nDWb2sJnlpjtGM/uVme0xsw0xZUmLycxyzOy3fvkqMytNUox3+b/r9Wb2P2Y2JtNijFn3TTNzZjYunTH2m3PuhPzBG0q4BjgZCAHrgPJBOvck4Cx/eSTeZODlwI+Am/zym4A7/eVyP74cYLofd5a/bjVwPt7sVU8C85Mc6/XAfwNP+M8zKkZgCfAP/nIIGJNJMeJNGbkVGOE/Xwr8fbpjBD4InAVsiClLWkzAPwE/95evAn6bpBg/AmT7y3dmYox+eQnekOXbgHHpjLHffyODdaKkB+69kU/FPF8ELEpTLI8BHwaqgUl+2SSgOl5s/h/N+f42G2PKPwvcn8S4pgArgIs4mtwzJkZgFF7itG7lmRRjZD7gIrwhsp/wE1TaYwRK6Zo4kxZTZBt/ORvvTkwbaIzd1n0C+E0mxgj8DjgDqOVock9bjP35OZGbZeJNwj15sIPwv2adCawCJjjndgH4j+P9zXqKdbK/3L08WX4CfBuInUA1k2I8GagHHvSbjn5hZvmZFKNz7h3g34DtwC7ggHPu6UyKMUYyY4ru45xrBw4AY5Mc7xfxarkZFaOZfRx4xzm3rtuqjImxL07k5J5wEu6UB2BWADwKfMM5d7C3TeOUuV7KkxHb5cAe59zrfd2lh1hS+T5n430lvs85dybQjNec0JN0vI+FwBV4X8NPAvLN7PO97dJDLOn8e+1PTCmN18xuBtqB3yQ436DGaGZ5wM3A9+Kt7uF8aXsfe3MiJ/cdeO1iEVOAnYN1cjMbjpfYf+Oc+71fvNvMJvnrJwF7EsS6w1/uXp4MFwIfN7Na4BHgIjP7rwyLcQewwzm3yn/+O7xkn0kxzgO2OufqnXNtwO+BCzIsxohkxhTdx8yygdFAQzKCNLOFwOXA55zfXpFBMZbhfZCv8/93pgBrzGxiBsXYJydyck/bJNz+lfBfAm85534cs+pxYKG/vBCvLT5SfpV/5Xw6MANY7X91PmRm5/nH/ELMPgPinFvknJvinCvFe2+edc59PsNifBeoM7PT/KKLgapMihGvOeY8M8vzj30x8FaGxRiRzJhij/VpvL+fAdc4zewS4Ebg4865lm6xpz1G59wbzrnxzrlS/39nB17niXczJcbjeTEn7A9wKV5PlRrg5kE87wfwvlqtB9b6P5fitaWtADb5j0Ux+9zsx1lNTC8JoALY4K/7D1JwsQWYy9ELqhkVIzAbqPTfyz8AhRkY4z8DG/3j/ydeb4m0xgg8jHcNoA0vAX0pmTEBucAyYDNeT5CTkxTjZrw26Mj/zc8zLcZu62vxL6imK8b+/mj4ARGRADqRm2VERKQHSu4iIgGk5C4iEkBK7iIiAaTkLiISQEruIiIBpOQuIhJA/x+VybPfaeAd2wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plot p values\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "neg_log_p_values = -1*np.log10(p_values_genome)\n",
    "\n",
    "\n",
    "plt.scatter(range(len(neg_log_p_values)), neg_log_p_values, s = 1)\n",
    "vert_line_pos = [int(e[3:]) for e in causal_variants] #[i for i, e in enumerate(['snp' + str(num1) for num1 in range(n_snps)]) if e in causal_variants]\n",
    "for position in vert_line_pos:\n",
    "    plt.axvline(x = position, color = 'b', label = 'axvline - full height')\n",
    "# plt.show()\n",
    "plt.savefig('simulated_p_values.png')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "24ffc6aa-7db5-479e-a75c-d1bce49fc446",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1234,\n",
       " 1236,\n",
       " 1634,\n",
       " 3477,\n",
       " 3533,\n",
       " 3996,\n",
       " 3997,\n",
       " 4473,\n",
       " 5865,\n",
       " 6501,\n",
       " 6692,\n",
       " 7979,\n",
       " 7982,\n",
       " 8849,\n",
       " 9487,\n",
       " 10693,\n",
       " 10696,\n",
       " 10945,\n",
       " 11124,\n",
       " 11144,\n",
       " 12127,\n",
       " 12908,\n",
       " 13041,\n",
       " 1234,\n",
       " 1236,\n",
       " 1634,\n",
       " 3477,\n",
       " 3533,\n",
       " 3996,\n",
       " 3997,\n",
       " 4473,\n",
       " 5865,\n",
       " 6501,\n",
       " 6692,\n",
       " 7979,\n",
       " 7982,\n",
       " 8849,\n",
       " 9487,\n",
       " 10693,\n",
       " 10696,\n",
       " 10945,\n",
       " 11124,\n",
       " 11144,\n",
       " 12127,\n",
       " 12908,\n",
       " 13041,\n",
       " 1234,\n",
       " 1236,\n",
       " 1634,\n",
       " 3477,\n",
       " 3533,\n",
       " 3996,\n",
       " 3997,\n",
       " 4473,\n",
       " 5865,\n",
       " 6501,\n",
       " 6692,\n",
       " 7979,\n",
       " 7982,\n",
       " 8849,\n",
       " 9487,\n",
       " 10693,\n",
       " 10696,\n",
       " 10945,\n",
       " 11124,\n",
       " 11144,\n",
       " 12127,\n",
       " 12908,\n",
       " 13041]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd195757-ee73-4edd-bd03-7c22856adad6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "318"
      ]
     },
     "execution_count": 293,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "len([0 for x in p_values_genome if math.isnan(x)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caccdaad-528f-43c7-b4b0-908b4ea7d6a9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
